src/Engine/Shaders/ParticleShader.cpp

```diff
@@ -0,0 +1,273 @@
+    glActiveTexture(GL_TEXTURE0 + cellDiffuseScalesTexture);
+    glBindTexture(GL_TEXTURE_1D, cellDiffuseScalesTexture);
+    glTexSubImage1D(GL_TEXTURE_1D, 0, 0, int(cellDiffuseScales.size()), GL_RED, GL_FLOAT, cellDiffuseScales.data());
+    glActiveTexture(GL_TEXTURE0 + cellSpecularScalesTexture);
+    glBindTexture(GL_TEXTURE_1D, cellSpecularScalesTexture);
+    glTexSubImage1D(GL_TEXTURE_1D, 0, 0, int(cellSpecularScales.size()), GL_RED, GL_FLOAT, cellSpecularScales.data());
+
+
+    for (ParticleComponent * pc : ParticleSystem::s_particleComponents) {
+        if (pc->getParticlePositions()->size() == 0) {
+            break;
+        }
+
+        // The reason there is a plus one is that there has to always be at least one element in RandomMs even if unused.
+        if (pc->RandomMs().size() > pc->MAX_ORIENTATIONS + 1) {
+            break;
+        }
+
+        /* Toon shading */
+        if (showToon) {
+            loadBool(getUniform("isToon"), true);
+        }
+        else {
+            loadBool(getUniform("isToon"), false);
+        }
+
+        bool r = (int)pc->RandomMs().size() != 1;
+        loadBool(getUniform("randomOrientation"), r);
+
+        /* Model matrix */
+        
+        /* Load Random Matrix Array */
+        loadMat4(getUniform("M"), pc->ModelMatrix());
+        loadMultiMat4(getUniform("randomMs"), &(pc->RandomMs()[0]), (int)pc->RandomMs().size());
+        
+        /* Load Normal matrix */
+        loadMat3(getUniform("N"), pc->NormalMatrix());
+        loadMultiMat3(getUniform("randomNs"), &(pc->RandomNs()[0]), (int)pc->RandomNs().size());
+
+
+
+        /* Bind materials */
+        loadFloat(getUniform("matAmbient"), pc->getModelTexture(0)->material.ambient);
+        loadVec3(getUniform("matDiffuse"), pc->getModelTexture(0)->material.diffuse);
+        loadVec3(getUniform("matSpecular"), pc->getModelTexture(0)->material.specular);
+        loadFloat(getUniform("shine"), pc->getModelTexture(0)->material.shineDamper);
+   
+        /* Load texture */
+        if(pc->getModelTexture(0)->texture && pc->getModelTexture(0)->texture->textureId != 0) {
+            loadBool(getUniform("usesTexture"), true);
+            loadInt(getUniform("textureImage"), pc->getModelTexture(0)->texture->textureId);
+            glActiveTexture(GL_TEXTURE0 + pc->getModelTexture(0)->texture->textureId);
+            glBindTexture(GL_TEXTURE_2D, pc->getModelTexture(0)->texture->textureId);
+        }
+        else {
+            loadBool(getUniform("usesTexture"), false);
+        }
+
+        /* Bind mesh */
+        glBindVertexArray(pc->getMesh(0)->vaoId);
+
+        /*Set Particle positions*/
+        int posPO = getAttribute("particleOffset");
+        unsigned int offsetBufId;
+        Vector<glm::vec3> *positions = pc->getParticlePositions();
+
+        /* Do all the buffer stuff */
+        glGenBuffers(1, &offsetBufId);
```

Comment:
We shouldn't be generating a new buffer per particle per frame!

Reply:
Unless I'm mustaken, we're not generating a buffer per particle. It's one buffer for all particles, that's why we're doing instanced rendering. The positions change for every frame, so this has to be updated.

Reply:
We should be binding updated data to an existing VBO on every frame, not generating an entirely new one

Reply:
Oh, I see. In that case I'll add that to ParticleComponent.

Reply:
yes it should be an existing vbo that is created in init and then updated using glBufferSubData